### ISSUE TITLE
CircularDependencyError when using ref.read

### ISSUE THREAD

**Issue by AhmedLSayed9** [rrousselGit/riverpod#1765]
**Describe the bug**

`CircularDependencyError` is thrown while I'm just interested in reading a value and not listening to changes. so my providers in fact doesn't depend on each others.

![Untitled Diagram drawio](https://user-images.githubusercontent.com/70890146/195904139-8bfe9254-a7ff-4e2c-8911-09af956edf7b.png)

The flow as the diagram showing is that my `dioProvider` has an `AuthInterceptor` which when calling post request it'll add the current `accessToken` to the request by calling `ref.read(accessTokenProvider)`. It's not interesting in updates from `accessTokenProvider`.

`AccessTokenNotifier` also needs to be able to use `dioProvider` to refresh the accessToken.

The problem is using `ref.read(accessTokenProvider)` inside the interceptor is considered a Circular Dependency.

**To Reproduce**
**Here's a simplified example to produce the issue:**

_Dio and the Interceptor:_
```
final dioProvider = Provider<Dio>((ref) {
  return Dio()
    ..interceptors.addAll([
      ref.watch(authInterceptorProvider),
    ]);
});

final authInterceptorProvider =
    Provider<AuthInterceptor>((ref) => AuthInterceptor(ref));

class AuthInterceptor extends QueuedInterceptor {
  AuthInterceptor(this.ref);

  final Ref ref;

  @override
  Future<void> onRequest(RequestOptions options, RequestInterceptorHandler handler) async {
    ref.read(accessTokenProvider);   //CircularDependencyError Exception

    return handler.next(options);
  }
}
```

_AccessToken Provider:_
```
final accessTokenProvider = StateNotifierProvider<AccessTokenNotifier, String?>((ref) {
  return AccessTokenNotifier(ref);
});

class AccessTokenNotifier extends StateNotifier<String?> {
  AccessTokenNotifier(this.ref) : super(null) {
    dio = ref.watch(dioProvider);
  }

  final Ref ref;
  late Dio dio;

  void refreshAccessToken() {
    //Call api
  }
}
```

_Usage:_
```
class TestScreen extends ConsumerWidget {
  const TestScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            ref.read(dioProvider).post('dummyInvalidPath');
          },
          child: const Text('add'),
        ),
      ),
    );
  }
}
```

**Expected behavior**
I understand that avoiding circular dependencies is necessary and `CircularDependencyError` is there for reasons. but shouldn't reading a provider not considered as a dependency?

In other words, we can solve the above issue by separating `accessTokenProvider` into 2 providers, one to hold the state and another to update it (as the following diagram):

![Untitled Diagram drawio](https://user-images.githubusercontent.com/70890146/195926961-fffa17d6-f1fb-4459-ace1-15846baae337.png)

But I don't get the advantage of doing that instead of allowing ref.read to work without considered as a CircularDependency.

**Comment by rrousselGit** [comment]
It **is** a circular dependency

Whether you want to listen to updates, you're still using the said provider. So although you don't subscribe to it, you do "depend" on it.


Although in some specific scenarios the error could possibly be emitted, the fact that there is no circular dependency in the graph of providers is a core design decision.
The entire framework is built around that rule, and changing it could have unexpected consequences.


Considering there's a way of achieving the same without a circular dependency, I don't think it's reasonable to remove the exception in the scenario you described.

**Comment by AhmedLSayed9** [comment]
I understand. Thanks for the clarification.
